### PR TITLE
[app] Add Dark mode

### PR DESCRIPTION
**Pull request by zerefwayne**
- Add option to switch between light/dark mode

https://user-images.githubusercontent.com/28949397/123642496-fda6b680-d840-11eb-92ab-79f1caa27f3a.mp4

To be merged after #3

### CODE REVIEW

File: diagnostics-app/src/app/app.component.html
Comment (MahmoudAGawad) [inline diagnostics-app/src/app/app.component.html]: Switch Theme might not be understood by all users. Let's be specific and call it something like "Enable Dark Mode" or just "Dark Mode".

Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.html]: Yeah this looks much better. 👍🏻 This button will eventually go away in the next PR though.

Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.html]: @MahmoudAGawad I've included a demo video in the PR description.

Comment (MahmoudAGawad) [reply diagnostics-app/src/app/app.component.html]: I cannot see the video

File: diagnostics-app/src/app/app.component.html
Comment (MahmoudAGawad) [inline diagnostics-app/src/app/app.component.html]: I'd just use one text, e.g. "Enable Dark Mode", and when the user toggles the button on, dark mode should be applied, another press would turn off the dark mode and toggle the button off. WDYT (What do you think)?

**Review by Lamzin — COMMENTED** [review]
I think it makes sense to use web APIs to detect default theme on the device. Also it will be good to save choice in some cookies, local storage or some other web storage.

btw, is there some existing package that handles themes?

File: diagnostics-app/src/app/app.component.html
Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.html]: My bad. Forgot to update the comment. It's visibile now! :)

File: diagnostics-app/src/app/app.component.html
Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.html]: This is to be be replaced by a sun/moon icon in the header in next PR. So we can leave this as is for now.

**Comment by zerefwayne** [comment]
@Lamzin I like these suggestions. 👍🏻 Fortunately, we don't need a package as browser gives inbuilt support to detect theme. There's a `prefers-color-scheme: 'dark'|'light'` available in Chrome which we can query. 

However, as discussed on the sync up, theming is a low priority issue. Is it okay if I take care of this once basic features are set up?

**Comment by Lamzin** [comment]
> Is it okay if I take care of this once basic features are set up?

Makes sense.

File: diagnostics-app/src/styles.css
Comment (zerefwayne) [inline diagnostics-app/src/styles.css:10]: auto lint

File: diagnostics-app/src/app/app.component.html
Comment (MahmoudAGawad) [reply diagnostics-app/src/app/app.component.html]: I see it now. But it is just changing the text color and not the background color. Is that the expected behavior?

File: diagnostics-app/src/app/app.component.html
Comment (MahmoudAGawad) [reply diagnostics-app/src/app/app.component.html]: Sounds good. Will that icon have a "hover over" text? This will make it more user-friendly.

**Comment by MahmoudAGawad** [comment]
> I think it makes sense to use web APIs to detect default theme on the device. Also it will be good to save choice in some cookies, local storage or some other web storage.
> 
> btw, is there some existing package that handles themes?

+1 but maybe in another PR (not a priority for now).

File: diagnostics-app/src/app/app.component.html
Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.html]: Yes I'll add that.

File: diagnostics-app/src/app/app.component.html
Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.html]: Yes there's no background as of now. The entire app is just the header component. The layout will be added in latter PRs. The color of background will be change then.

File: diagnostics-app/angular.json
Comment (MahmoudAGawad) [inline diagnostics-app/angular.json]: Is this related to the PR goal?

Comment (zerefwayne) [reply diagnostics-app/angular.json]: My bad. This should've been in the Fix CI. I'll remove this from here and raise a new PR for fixing the hook variable name and adding this variable.

Comment (zerefwayne) [reply diagnostics-app/angular.json]: Removed this from here. 👍🏻

File: diagnostics-app/src/app/app.component.spec.ts
Comment (MahmoudAGawad) [inline diagnostics-app/src/app/app.component.spec.ts:49]: Let's also check the theme value was changed:
```
    // css class 'light' should be added
    expect(component.theme).toBe('light');
    expect(component.cssClass).toBe('light');
```

Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.spec.ts:49]: I had this doubt while writing the test. Since I have already tested the toggleTheme function, wouldn't this become redundant?

File: diagnostics-app/src/app/app.component.ts
Comment (MahmoudAGawad) [inline diagnostics-app/src/app/app.component.ts]: these variables should be private.

Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.ts]: Okay, I have added getters for them.

File: diagnostics-app/src/app/app.component.spec.ts
Comment (MahmoudAGawad) [reply diagnostics-app/src/app/app.component.spec.ts:49]: Correct.

The previous test is a unit_test, while this one is integration/component test (that's clicking on the button triggers the appropriate logic, and changes the system state as expected). Generally, we need both. Maybe we can split the two types of tests in different files? I am not familiar with Angular, what is the best practice here?

Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.spec.ts:49]: Integration tests in terms of Angular are where two or more components are interacting together. In normal contexts this would have been an integration test. But since the button is part of the same component, it is considered to be a unit test. If this button was a different component, it will be called an integration test.

Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.spec.ts:49]: > In Angular, the difference between unit and integration tests is sometimes subtle. A unit test typically focusses on a single Angular Component, Directive, Service, Pipe, etc. Dependencies are replaced with fakes. An integration test spans one Component together with its children and possibly connected Services as well. It is also possible to write a test that integrates all parts of an Angular Module.

From, https://testing-angular.com/testing-principles/#distribution-of-testing-efforts

Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.spec.ts:49]: This is a relevant example of how we can structure the tests: [Example](https://github.com/9elements/angular-workshop/blob/main/src/app/components/service-counter/service-counter.component.spec.ts)

Contrary to this, when I write integration tests, we can keep them in separate files such as:
`app.component.unit.spec.ts` and `app.component.it.spec.ts`

But I think we should keep them in a single `spec.ts` file as there are already so many files in the project structure. We can have separate describe blocks for both of them just like the example.

**Review by Lamzin — APPROVED** [review]
LGTM, i'll defer review for Mahmoud. Thanks!

File: diagnostics-app/src/app/app.component.spec.ts
Comment (MahmoudAGawad) [reply diagnostics-app/src/app/app.component.spec.ts:49]: Thanks for sharing. I am convinced that both are unit tests but since there is redundancy, in my opinion, we should keep the second test and remove the first one. This entails checking theme and cssClass fields after clicking the button.

File: diagnostics-app/src/app/app.component.html
Comment (MahmoudAGawad) [reply diagnostics-app/src/app/app.component.html]: if this is the case, then let's wait until the layout PR is merged and then merge this one. After this PR is merged, change theme should be working as expected.

Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.html]: No actually this PR is necessary for the next ones. We can merge this for now 👍 The dark theme setup we do here, and the header component in the next PR, both are necessary for the app layout one.

File: diagnostics-app/src/app/app.component.spec.ts
Comment (zerefwayne) [reply diagnostics-app/src/app/app.component.spec.ts:49]: Sure, I combined the tests into one! 👍🏻

File: diagnostics-app/src/app/app.component.html
Comment (MahmoudAGawad) [reply diagnostics-app/src/app/app.component.html]: Offline discussion: Decided to add a background color that will be flipped with dark mode button toggle.

**Review by MahmoudAGawad — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**
